### PR TITLE
script: previous_release.sh rewritten in python

### LINKED ISSUE #18132: contrib/devtools/previous_release.sh should be rewritten to python
Most of our test scripts are written in python. We don't have enough reviewers for bash scripts and they tend to be clumsy anyway. Especially when it comes to argument parsing.

Thus, contrib/devtools/previous_release.sh should be rewritten in python

#### Useful skills:

Experience with Bash and Python

#### Want to work on this issue?

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by bliotti**
Closes #18132

Added functionality:
1) checks file hash before untarring when using the binary download option

### CODE REVIEW

**Review by Rickyricky1987 — APPROVED** [review]

File: contrib/devtools/previous-release.py
Comment (Sjors) [inline contrib/devtools/previous-release.py:72]: `encoding="utf-8"` shouldn't be necessary and results in `ValueError: binary mode doesn't take an encoding argument`.

File: contrib/devtools/previous-release.py
Comment (Sjors) [inline contrib/devtools/previous-release.py:147]: This script should not build QT (see below), so this can be dropped.

File: contrib/devtools/previous-release.py
Comment (Sjors) [inline contrib/devtools/previous-release.py:122]: This should always be set.

File: contrib/devtools/previous-release.py
Comment (Sjors) [inline contrib/devtools/previous-release.py:185]: These flags should always be set.

**Review by Sjors — CHANGES_REQUESTED** [review]
Concept ACK, thanks for working on this. 6e4a07c83ccc099e3b60d644d58e3f6558dca021 works on macOS 10.15.4 for the standard set of binaries we download, release candidates, and for compiling v0.20.0.

Can you squash your commits?

You can also remove `previous_release.sh` and switch over to this Python version (in a separate commit).

Are you using Python 3.5.6? (pyenv should do that automatically)

If you check the checksums then it makes sense to also check their PGP signature. Alternatively we might as well hard code the checksum for each `tar.gz` release in the source code, here. That way Travis might catch someone messing with `bitcoincore.org`.

I initially added `-f   Configure for functional tests` as a way to opt-out of building QT. Looking at it again, I think we shouldn't build QT period. So I added inline suggestions to drop QT everywhere.

**Comment by practicalswift** [comment]
Concept ACK

Nice first-time contribution @bliotti - welcome as a contributor! :)

**Comment by bliotti** [comment]
Thanks, I have always wanted to contribute! I will follow up with the mentioned suggestions. :)

File: contrib/devtools/previous-release.py
Comment (fanquake) [inline contrib/devtools/previous-release.py:34]: You could add return type annotations to function definitions. i.e:
```python
def download_binary(tag, args) -> int:
```

File: contrib/devtools/previous-release.py
Comment (fanquake) [inline contrib/devtools/previous-release.py:104]: Could use `subprocess.run(..., check=True)` here and where you're using `subprocess.call()` etc; as opposed to the older subprocess API.

**Comment by bliotti** [comment]
> Concept ACK, thanks for working on this. 6e4a07c83ccc099e3b60d644d58e3f6558dca021 works on macOS 10.15.4 for the standard set of binaries we download, release candidates, and for compiling v0.20.0.
> 
> 
> 
> Can you squash your commits?
> 
> 
> 
> You can also remove `previous_release.sh` and switch over to this Python version (in a separate commit).
> 
> 
> 
> Are you using Python 3.5.6? (pyenv should do that automatically)
> 
> 
> 
> If you check the checksums then it makes sense to also check their PGP signature. Alternatively we might as well hard code the checksum for each `tar.gz` release in the source code, here. That way Travis might catch someone messing with `bitcoincore.org`.
> 
> 
> 
> I initially added `-f   Configure for functional tests` as a way to opt-out of building QT. Looking at it again, I think we shouldn't build QT period. So I added inline suggestions to drop QT everywhere.


Yes I will squash my commits. 

Just to confirm, I need to squash all these commits into **one** commit that encompasses the addition of the py script, and then... push up one more commit for the removal of the bash script.

Most of the suggestions have been incorporated I just need to add the PGP signature check and I should have it up soon.

File: contrib/devtools/previous-release.py
Comment (bliotti) [reply contrib/devtools/previous-release.py:139]: I changed the other `subprocess.call` methods to `subprocess.run` but when I used `run` for this line it hung up the script. I had to exit out of the tag list to continue the script. However `check_output` worked.

File: contrib/devtools/previous-release.py
Comment (Sjors) [inline contrib/devtools/previous-release.py:102]: I don't think you should mess with the developers PGP keys.

File: contrib/devtools/previous-release.py
Comment (Sjors) [inline contrib/devtools/previous-release.py:105]: `text=True` causes an error for me: `TypeError: __init__() got an unexpected keyword argument 'text'`

Also:
```
  File "/Users/sjors/.pyenv/versions/3.5.6/lib/python3.5/subprocess.py", line 383, in run
    with Popen(*popenargs, **kwargs) as process:
TypeError: __init__() got an unexpected keyword argument 'capture_output'
```

**Review by Sjors — COMMENTED** [review]
I restarted Travis for an unrelated failure (details in #19221) 

It's probably better to move `closes #18132` into the commit message body instead of the subject (needs a blank line). Title could be: `test: switch usage of previous_release.sh to .py `. Right now the changes are mixed between both commits; a single commit is fine too if it's tricky to untangle.

Let's not download and add PGP keys on behalf on the user. You could check if the key is present and fail otherwise.

**Comment by bliotti** [comment]
Will jump on this tonight.

**Comment by bliotti** [comment]
> It's probably better to move `closes #18132` into the commit message body instead of the subject (needs a blank line). Title could be: `test: switch usage of previous_release.sh to .py `. Right now the changes are mixed between both commits; a single commit is fine too if it's tricky to untangle.

did you mean `test: switch usage of` or `script: switch usage of....`

**Comment by fjahr** [comment]
Concept ACK

Great first contribution @bliotti! I will take a closer look one you have squashed the commits. You can make yourself a co-author as well, I think. And I think @Sjors mistyped, it should be `script: ...`.

**Comment by bliotti** [comment]
Not sure why AppVevyor test failed?

**Comment by bliotti** [comment]
@fjahr did you want to review it now? Thanks!

**Review by fjahr — APPROVED** [review]
tested ACK 60faac3351866730af783da3097a84b906ad8175

However, the docs should be updated. CI failure seems to be unrelated. I could restart Appveyor but it seems I can't restart that Travis build.

Also, the formatting changes in the second commit make this a longer review than necessary and I think many of them are not worth the noise they create. In some cases, the original authors even made an effort to line up comments etc. I don't use that style but it's not necessary to remove it unless you are editing that particular part of the file. For me, it's ok to keep them but other reviewers might disagree. If you keep them they should be moved into their own commit so that their current commit does what the commit message says: `script: removed previous_release.sh`. I would suggest to make it:
```
script: switch previous_release from .sh to .py version

Remove previous_release.sh
```
(the switch is more significant than the removal of the file). 

And then the formatting changes move to `refactor: Clean up functional compat test formatting`.

EDIT: actually, the first commit message should be changed as well then because it doesn't do the switch now. Should be just something like `script: Add previous_release.py`.

File: test/functional/wallet_upgradewallet.py
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:93]: I don't really like the formatting changes here. This makes it harder to read, imo.

Maybe leave them out for now?

File: test/functional/wallet_upgradewallet.py
Comment (bliotti) [reply test/functional/wallet_upgradewallet.py:31]: This was unintentional. This was from an auto formatting tool I was not paying attention to. It will be fixed and realigned.

File: test/functional/wallet_upgradewallet.py
Comment (bliotti) [reply test/functional/wallet_upgradewallet.py:93]: All of this formatting was unintentional. I didn't notice it auto formatted the file when I edited the one line above. It will be put back.

File: test/functional/feature_backwards_compatibility.py
Comment (bliotti) [reply test/functional/feature_backwards_compatibility.py:9]: good catch.

File: contrib/devtools/previous-release.py
Comment (bliotti) [reply contrib/devtools/previous-release.py:98]: What are you thinking? Maybe there's no need for this check? I am ok with losing it.

Comment (fjahr) [reply contrib/devtools/previous-release.py:98]: sorry, I was just trying to be be funny. All I wanted to say was there is a typo: should probably be `signingKey` not `singingKey` :)

Comment (bliotti) [reply contrib/devtools/previous-release.py:98]: ohh lol... Totally went over my head. 😑 Thanks 👍

File: contrib/devtools/previous_release.py
Comment (MarcoFalke) [inline contrib/devtools/previous_release.py:229]: extras means tags?

If yes, this can be a positional argument with the `+` specifier (one or more instances)

File: contrib/devtools/previous_release.py
Comment (MarcoFalke) [inline contrib/devtools/previous_release.py:21]: I think python prints the help by itself already, no?

**Comment by fjahr** [comment]
I think for the "co-authored-by:" to work as intended it needs to be at the end of the commit message. You can check that it shows both profile pictures and in Github it says 'bliotti and bboot committed'. You also only need to list the additional co-authors, not the author of the commit itself (yourself in this case).

See [`4950e5d` (#19055)](https://github.com/bitcoin/bitcoin/pull/19055/commits/4950e5d36dfa4219234f5e176bc4c2dabd44a397) for example.

File: contrib/devtools/previous_release.py
Comment (bliotti) [reply contrib/devtools/previous_release.py]: removed usage in 9c34aff

File: contrib/devtools/previous_release.py
Comment (bliotti) [reply contrib/devtools/previous_release.py]: added in 9c34aff

**Comment by fjahr** [comment]
re-ACK 9c34aff39309b8adc99d347e07b6ddb5366498e9

**Comment by bliotti** [comment]
thanks for the help @fjahr :+1:

**Comment by Sjors** [comment]
tACK 9c34aff39309b8adc99d347e07b6ddb5366498e9

**Comment by bliotti** [comment]
Is there anything I need to do now?

**Comment by fjahr** [comment]
> Is there anything I need to do now?

Not really, but you could respectfully ping someone to get another review/ACK. With three ACKs this should be ready for merge and if maintainers don't take notice then you could post a message in IRC that this might be RFM after a few days.

**Comment by jonatack** [comment]
> Is there anything I need to do now?

@bliotti welcome! May I suggest this section (particularly the last item) https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#finding-reviewers, followed by these steps:

- Participate in https://bitcoincore.reviews and check out the resources listed at the bottom of the home page
- Help out reviewing some of the hundreds of open issues and PRs that are waiting for testing or review, for instance among these by your kind reviewers :slightly_smiling_face:  https://github.com/bitcoin/bitcoin/pulls/fjahr and https://github.com/bitcoin/bitcoin/pulls/Sjors

**Comment by emsheets** [comment]
Concept ACK, will test later today if test ACKs still needed at that time. Great stuff, @bliotti!!

**Comment by MarcoFalke** [comment]
I did some minor cleanups in #19560

**Comment by bliotti** [comment]
> I did some minor cleanups in #19560

Thanks! @MarcoFalke

**Comment by hebasto** [comment]
> Alternatively we might as well hard code the checksum for each `tar.gz` release in the source code, here. That way Travis might catch someone messing with `bitcoincore.org`.

Please see #19813.

**Pull request merged**
